src/store.c

```diff
@@ -339,144 +373,255 @@ static void writer_close(
     close(store->fd);
 }
 
+static void init_store_offsets(struct rill_store* store, uint64_t pairs)
+{
+    store->head->index_b_off = store->head->index_a_off + indexer_cap(pairs);
+    store->head->data_a_off = store->head->index_b_off + indexer_cap(pairs);
+
+    store->index_a = (void *) ((uintptr_t) store->vma + store->head->index_a_off);
+    store->index_b = (void *) ((uintptr_t) store->vma + store->head->index_b_off);
+    store->data_a = (void *) ((uintptr_t) store->vma + store->head->data_a_off);
+}
+
+static void prepare_col_b_offsets(
+    struct rill_store* store, struct encoder* coder_a)
+{
+    store->head->data_b_off = store->head->data_a_off + coder_off(coder_a);
+    store->data_b = (void *) ((uintptr_t) store->vma + store->head->data_b_off);
+}
+
 bool rill_store_write(
         const char *file,
-        rill_ts_t ts, size_t quant,
+        rill_ts_t ts,
+        size_t quant,
         struct rill_pairs *pairs)
 {
     rill_pairs_compact(pairs);
     if (!pairs->len) return true;
 
     struct vals *vals = vals_from_pairs(pairs);
     if (!vals) goto fail_vals;
+    struct vals *invert_vals = keys_from_pairs(pairs); // TODO fix fail cleanup
```

Comment:
I see todos

Also bzz bzz...

Reply:
I missed this. I'll fix this, sorry about that.

Reply:
Instead of having 2 functions for each column, I would just add the extra parameter to grab a specific column which is more consistent. note that `vals_` here is meant as a namespace for all `vals_` related functions which is why having a `keys_` prefix is problematic.

Reply:
fixorzed, though now it's called `vals_cols_from_pairs` instead.